elasticsearch/src/main/java/org/apache/calcite/adapter/elasticsearch/ElasticsearchAggregate.java

```diff
@@ -0,0 +1,170 @@
+      if (aggCall.isDistinct()) {
+        throw new InvalidRelException("distinct aggregation not supported");
+      }
+
+      SqlKind kind = aggCall.getAggregation().getKind();
+      if (!SUPPORTED_AGGREGATIONS.contains(kind)) {
+        final String message = String.format(Locale.ROOT,
+            "Aggregation %s not supported (use one of %s)", kind, SUPPORTED_AGGREGATIONS);
+        throw new InvalidRelException(message);
+      }
+    }
+
+    if (getGroupType() != Group.SIMPLE) {
+      final String message = String.format(Locale.ROOT, "Only %s grouping is supported. "
+              + "Yours is %s", Group.SIMPLE, getGroupType());
+      throw new InvalidRelException(message);
+    }
+
+  }
+
+  @Override public Aggregate copy(RelTraitSet traitSet, RelNode input, boolean indicator,
+      ImmutableBitSet groupSet, List<ImmutableBitSet> groupSets,
+      List<AggregateCall> aggCalls) {
+    try {
+      return new ElasticsearchAggregate(getCluster(), traitSet, input,
+          indicator, groupSet, groupSets,
+          aggCalls);
+    } catch (InvalidRelException e) {
+      throw new AssertionError(e);
+    }
+  }
+
+  @Override public RelOptCost computeSelfCost(RelOptPlanner planner, RelMetadataQuery mq) {
+    return super.computeSelfCost(planner, mq).multiplyBy(0.1);
+  }
+
+  @Override public void implement(Implementor implementor) {
+    implementor.visitChild(0, getInput());
+    List<String> inputFields = fieldNames(getInput().getRowType());
+
+    if (groupSet.cardinality() == 1) {
+      final String groupByFieldName = inputFields.get(groupSet.nth(0));
+      implementor.addGroupBy(groupByFieldName);
+    } else {
+      for (int group : groupSet) {
+        implementor.addGroupBy(inputFields.get(group));
+      }
+    }
+
+    for (AggregateCall aggCall : aggCalls) {
+      List<String> names = new ArrayList<>();
+      for (int i : aggCall.getArgList()) {
+        names.add(inputFields.get(i));
+      }
+
+      final String name = names.isEmpty() ? ElasticsearchConstants.ID : names.get(0);
+
+      String op = String.format(Locale.ROOT, "\"%s\":{\"field\": \"%s\"}",
+          toElasticAggregate(aggCall),
+          name);
+
+      implementor.addAggregation(aggCall.getName(), op);
+    }
+  }
+
+  /**
+   * Most of the aggregations can be retrieved with single
+   * <a href="https://www.elastic.co/guide/en/elasticsearch/reference/current/search-aggregations-metrics-stats-aggregation.html">stats</a>
+   * function. But currently only one-to-one mapping is supported between sql agg and elastic
+   * aggregation.
+   */
+  private String toElasticAggregate(AggregateCall call) {
+    SqlKind kind = call.getAggregation().getKind();
+    switch (kind) {
+    case COUNT:
+      return call.isApproximate() ? "cardinality" : "value_count";
+    case SUM:
+      return "sum";
+    case MIN:
+      return "min";
+    case MAX:
+      return "max";
+    case AVG:
+      return "avg";
+    default:
```

Comment:
SUM0 is not going to be supported?

Reply:
Should `SUM0` be identical to `SUM` in ES context ? I didn't see `SUM0` being used except for jdbc adapter.

Reply:
Hi @hsyuan I would like to close all open questions with this PR. Can you please let me know how exactly `SUM0` should be implemented ?

Reply:
@asereda-gs Sorry for my late response. It is fine to not support SUM0 if SUM0 is identical to SUM in ES context. This question can be closed. Thanks!